HyperVExtension/src/HyperVExtension/Models/VirtualMachineCreation/VMGalleryVMCreationOperation.cs

```diff
@@ -0,0 +1,247 @@
+
+    private readonly IArchiveProviderFactory _archiveProviderFactory;
+
+    private readonly IHyperVManager _hyperVManager;
+
+    private readonly IDownloaderService _downloaderService;
+
+    private readonly string _tempFolderSaveLocation = Path.GetTempPath();
+
+    private readonly IStringResource _stringResource;
+
+    private readonly IVMGalleryService _vmGalleryService;
+
+    private readonly VMGalleryCreationUserInput _userInputParameters;
+
+    public CancellationTokenSource CancellationTokenSource { get; private set; } = new();
+
+    private readonly object _lock = new();
+
+    public bool IsOperationInProgress { get; private set; }
+
+    public bool IsOperationCompleted { get; private set; }
+
+    public CreateComputeSystemResult? ComputeSystemResult { get; private set; }
+
+    public StorageFile? ArchivedFile { get; private set; }
+
+    public VMGalleryImage Image { get; private set; } = new();
+
+    public event TypedEventHandler<ICreateComputeSystemOperation, CreateComputeSystemActionRequiredEventArgs>? ActionRequired = (s, e) => { };
+
+    public event TypedEventHandler<ICreateComputeSystemOperation, CreateComputeSystemProgressEventArgs>? Progress;
+
+    public VMGalleryVMCreationOperation(
+        IStringResource stringResource,
+        IVMGalleryService vmGalleryService,
+        IDownloaderService downloaderService,
+        IArchiveProviderFactory archiveProviderFactory,
+        IHyperVManager hyperVManager,
+        VMGalleryCreationUserInput parameters)
+    {
+        _stringResource = stringResource;
+        _vmGalleryService = vmGalleryService;
+        _userInputParameters = parameters;
+        _archiveProviderFactory = archiveProviderFactory;
+        _hyperVManager = hyperVManager;
+        _downloaderService = downloaderService;
+    }
+
+    /// <summary>
+    /// Reports the progress of an operation.
+    /// </summary>
+    /// <param name="value">The archive extraction operation returned by the progress handler which extracts the archive file</param>
+    public void Report(IOperationReport value)
+    {
+        var displayText = Image.Name;
+
+        if (value.ReportKind == ReportKind.ArchiveExtraction)
+        {
+            displayText = $"{ArchivedFile!.Name} ({Image.Name})";
+        }
+
+        UpdateProgress(value, value.LocalizationKey, displayText);
+    }
+
+    /// <summary>
+    /// Starts the VM gallery operation.
+    /// </summary>
+    /// <returns>A result that contains information on whether the operation succeeded or failed</returns>
+    public IAsyncOperation<CreateComputeSystemResult?> StartAsync()
+    {
+        return Task.Run(async () =>
+        {
+            try
+            {
+                lock (_lock)
+                {
+                    if (IsOperationInProgress)
+                    {
+                        var exception = new OperationInProgressException(_stringResource);
```

Comment:
Note for future changes: We could consider returning the current operation here if it's already in progress. It seems inconsistent that we throw an exception if the operation is in progress but return the operation result if it's completed (someone could expect that we'd start it again, for example instead of returning previous result).

Reply:
Thats fair, I'll keep that in mind for the future